### PR TITLE
Cartesian Builder

### PR DESCRIPTION
**Pull request by raulraja**
Cartesian Builder for Applicative instances

Pain Driven Development :tm:

### CODE REVIEW

**Comment by raulraja** [comment]
Open to anyone that wants to take a look, achieved syntax is in the test

File: katz/src/main/kotlin/katz/typeclasses/Applicative.kt [hunk 1]
```diff
@@ -16,17 +16,57 @@
 
 package katz
 
+
```
Comment (ffgiraldez) [inline katz/src/main/kotlin/katz/typeclasses/Applicative.kt]: remove empty line

File: katz/src/main/kotlin/katz/typeclasses/Applicative.kt [hunk 2]
```diff
@@ -16,17 +16,57 @@
...
-            ap(fb, map(fa) { a: A -> { b: B -> Pair(a, b) } })
+    fun <A, B> product(fa: HK<F, A>, fb: HK<F, B>): HK<F, Tuple2<A, B>> =
+            ap(fb, map(fa) { a: A -> { b: B -> Tuple2(a, b) } })
 
     override fun <A, B> map(fa: HK<F, A>, f: (A) -> B): HK<F, B> = ap(fa, pure(f))
 
-    fun <A, B, Z> map2(fa: HK<F, A>, fb: HK<F, B>, f: (Pair<A, B>) -> Z): HK<F, Z> =
+    fun <A, B, Z> map2(fa: HK<F, A>, fb: HK<F, B>, f: (Tuple2<A, B>) -> Z): HK<F, Z> =
             map(product(fa, fb), f)
+
```
Comment (ffgiraldez) [inline katz/src/main/kotlin/katz/typeclasses/Applicative.kt]: same here

**Comment by raulraja** [comment]
@FineCinnamon/maintainers this is ready for merge, any objections?

File: katz/src/main/kotlin/katz/typeclasses/Applicative.kt [hunk 3]
```diff
@@ -42,31 +41,257 @@ data class Tuple5<out A, out B, out C, out D, out E>(val a: A, val b: B, val c:
 data class Tuple6<out A, out B, out C, out D, out E, out F>(val a: A, val b: B, val c: C, val d: D, val e: E, val f: F)
 data class Tuple7<out A, out B, out C, out D, out E, out F, out G>(val a: A, val b: B, val c: C, val d: D, val e: E, val f: F, val g: G)
 data class Tuple8<out A, out B, out C, out D, out E, out F, out G, out H>(val a: A, val b: B, val c: C, val d: D, val e: E, val f: F, val g: G, val h: H)
+data class Tuple9<out A, out B, out C, out D, out E, out F, out G, out H, out I>(val a: A, val b: B, val c: C, val d: D, val e: E, val f: F, val g: G, val h: H, val i: I)
+data class Tuple10<out A, out B, out C, out D, out E, out F, out G, out H, out I, out J>(val a: A, val b: B, val c: C, val d: D, val e: E, val f: F, val g: G, val h: H, val i: I, val j: J)
 
 
-fun <F, A, B> HK<F, A>.product(AP: Applicative<F>, other: HK<F, B>): HK<F, Tuple2<A, B>> =
+fun <μ, A, Z> HK<μ, A>.product(AP: Applicative<μ>, other: HK<μ, Z>): HK<μ, Tuple2<A, Z>> =
```
Comment (pakoito) [inline katz/src/main/kotlin/katz/typeclasses/Applicative.kt]: I'd avoid using μ, because the Android JVM isn't too happy about non-regular characters.

**Pull request merged**